src/main/java/com/artipie/npm/proxy/NpmProxy.java

```diff
@@ -0,0 +1,361 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 artipie
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.npm.proxy;
+
+import com.artipie.asto.Concatenation;
+import com.artipie.asto.Content;
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.fs.RxFile;
+import com.artipie.asto.rx.RxStorage;
+import com.artipie.asto.rx.RxStorageWrapper;
+import com.artipie.npm.proxy.json.PackageJsonTransformer;
+import com.artipie.npm.proxy.model.NpmAsset;
+import com.artipie.npm.proxy.model.NpmPackage;
+import com.jcabi.log.Logger;
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import io.reactivex.MaybeSource;
+import io.reactivex.Single;
+import io.vertx.core.file.OpenOptions;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.RequestOptions;
+import io.vertx.ext.web.client.WebClientOptions;
+import io.vertx.reactivex.core.Vertx;
+import io.vertx.reactivex.core.buffer.Buffer;
+import io.vertx.reactivex.core.file.AsyncFile;
+import io.vertx.reactivex.ext.web.client.HttpResponse;
+import io.vertx.reactivex.ext.web.client.WebClient;
+import io.vertx.reactivex.ext.web.codec.BodyCodec;
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import net.minidev.json.JSONObject;
+import net.minidev.json.JSONValue;
+import net.minidev.json.parser.JSONParser;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * NPM Proxy.
+ * @since 0.1
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public class NpmProxy {
```

Comment:
@brastak this class also has to be refactored and divided into smaller classes to be easier to read and understand, easier to test, easier to maintain.

Reply:
Is it really hard to read, understand, test and maintain now? It has 2 public method that implements **all** functionality of this adapter in one place. The alternative you are talking about is:
1) leave this class because outbound http connections should be managed from one place;
2) add 2 new classed each of them implementing operations with own object type;
3) add one more abstract class that should be the parent of the 2 classes from the p.2 because of they uses the same logic when create requests.
Will it really simpler than the current state? When the code will be complicated (and it will be because of number of the features is planned), then this class becomes too big and will be divided.

Reply:
@brastak hm, the fact that there is a need to suppress `ClassDataAbstractionCouplingCheck` is the first not made up by me prove that this class is overloaded with functionality. 
Now let's take a look at the private methods and what fields they use:
1. `tryLocal` works only with `storage`
2. `readPackageFromStorage` is called from `tryLocal` and also works only with `storage`
3. `readAssetFromStorage` also works only with `storage`
4. `buildRequest` works only with `settings`

So first 3 private methods depend on `storage` only, their main work is to read smth from it. So why not to extract 3 methods into separate class depending on `storage`? This new class will be smaller, easier to test and maintain and `NpmProxy` become smaller too.

Reply:
I have added puzzle for this